Feature/twbs4theme

stroebjo opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks backward-compatibility?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #573, #568
| Updated README/docs?   | ❌
| Added CHANGELOG entry?   | ❌

As discusses in #573 I looked over the Bootstrap 4 theme and updated it according to the latest Bootstrap documentation. 

**Theme options**

| Option           | Default | Description
| ------------- | ------ | -----
| `input_size`    | `normal` | Changes `input` size, can be `normal`, `small` or `large`
| `custom_forms`  | `false` | Apply Bootstrap custom forms
| `object_indent`  | `true` | Indent nested object elements (use nested `.card` layout)
| `object_background`  | `bg-light` | Bootstrap 4 card background [modifier class](https://getbootstrap.com/docs/4.1/getting-started/introduction/) 
| `object_text`  | `''` | Bootstrap 4 card text color modifier class [modifier class](https://getbootstrap.com/docs/4.1/getting-started/introduction/) 
| `table_border`  | `false` | Add border to array "table" row and cells
| `table_zebrastyle`  | `false` | Add "zebra style" to array "table" rows
| `tooltip`  | `bootstrap` | how to display tooltips (infoText). Can be `browser` for native `[title]`, `css` for simple CSS Styling, or `bootstrap` for TWBS/Popper.js handling


A few remarks / questions:

- `getButtonHolder()` is the wrapper function for button groups on the panel header, as well as for the controls on "multiple item" things. In the panel header case it wraps each button individually, on the other case it wraps all buttons, as expected. Because of this the TWBS4 class `btn-group` can not be used, since this would lead to cut of corners of a single button in a button group. Maybe this can be unified? Probably would need work in all themes though?
- `getMultiCheckboxHolder()` has no `compact` Parameter like `getFormRadioControl()`, so I can't really distinguish on when to use the compact (horizontal) view and when not.
- I found the panel layouts a bit limiting. For TWBS4 I thought about using the [card](https://getbootstrap.com/docs/4.1/components/card/#header-and-footer) component, but I coudn't quite get it working after adding additional `div`-layers. 
- I added `for`/`id` attributes to select the input by clicking on the `label` (89b3652 - feels a bit hacky ;)). But this doesn't seem to catch `[type=checkbox|file]`? In combination with the custom forms of Bootstrap this is a problem, since the checkbox ist not otherwise selectable than clicking on the label. Maybe this can be done in the core? Would also benefit all themes then.

Let me know if I can somehow improve the pull request!

germanbisurgi commented:
this is great :)
I addressed the first point in #587. Buttons were placed badly. For example "btn-group" with only one button breaking border radius. It still not perfect and must to be updated but i did not want to make to much changes in one PR. What i did though was to write extensive test basis for future themes upgrades. 

One of the file i updated in  #587 is the "bootstrap 4" theme ("getButton" holder and "getButton") so a little merge conflict could happen but very easi to resolve

pmk65 commented:
The 2nd is because I haven't gotten around to refactoring the multiselect checkboxes (See ToDo at #456 ).

pmk65 commented:
> 
> I addressed the first point in #587. Buttons were placed badly. For example "btn-group" with only one button breaking border radius. It still not perfect and must to be updated but i did not want to make to much changes in one PR. What i did though was to write extensive test basis for future themes upgrades.
> 

The problem is that rounded corners of buttons inside button groups is controlled by CSS, using `:first-child` / `:first-child` selectors. 
So when one of the edge buttons inside the button group is hidden, it will still apply the rounded corners to the hidden button and not the first/last visible button. 
Example: https://jsfiddle.net/bx8u4wsL/

stroebjo commented on src/themes/bootstrap4.js:
```diff
@@ -3,114 +3,372 @@ import { AbstractTheme } from '../theme'
     return group
   },
+
+  getInfoButton: function (text) {
```
When using the Bootstrap theme I would expect every element to conform to framework styling - in the case of the tooltip this requires the `tooltip()` call (this also checks for alignment of the tooltip on the edges of the browser view top/bottom etc).

But I agree that that not everyone might use the Bootstrap JavaScript stuff. I added an option to use your suggested CSS styling.

stroebjo commented:
@pmk65 thanks for your feedback! 

I added an option for `object_indent` (in case of the Bootstrap card display I don't think it's useful to split `border`/`margin` into two options). I also added table options like in the Specter theme.

I fixed the styling for the Select2 und Selectize plugins. But the size matches only `input_size=normal`, also if someone changed the sizes variables in his Bootstrap setup this can not be reflected in this JSON-Editor theme. In this cases one would need to add his own style fixes.

Regarding the tooltip I commented above.

pmk65 commented on src/themes/bootstrap4.js:
```diff
@@ -3,114 +3,372 @@ import { AbstractTheme } from '../theme'
-      input.controlgroup.style.marginBottom = 0
+
+    // set id/for
+    // is not working for: [type=file], [type=checkbox]
```
This makes sense, as the checkbox and file editors have not been refactored yet. So they might not be theme friendly like the other editors/types.

pmk65 commented:
@stroebjo Looks good. I'll update the Bootstrap4 testcase with toggle buttons for the options.

One more thing that needs fixing is the fontsize in the Properties Modal box, which is way too big due to it inheriting the parent H4 fontsize.
But don't try fixing it yet as #587 makes changes to the layout. So we need to get that one merged first.

pmk65 commented:
Noticed a weird behavior. 
1) Load the "Button Editor" example on the Playground, then switch to the new Bootstrap4 Theme.
2) Press the "Generate" button.

For me the content on the 1st Tab "Personal info" is hidden. Only when I click the Tab (or switch between that and another Tab), the content become visible.

Also the buttons doesn't respect the alignment parameter. (Everything is left aligned)

stroebjo commented:
I added the alignment options for the buttons (there is a small offset on the right side, since I added my workaround for the spacing + `btn-group` issue. I think it would be best to look at this, after those are resolved?).

Regarding the tabs I'm a bit confused. This doesn't happen in the Bootstrap 4 Example ("Array with format: tabs-top"), but also happens for "Object w. format: categories". I looked at the tab rendering in  `editors/object.js` - but that was a bit too much…

- when I look in the `row` parameter in `markTabActive(row)` I noticed that for the initial rendering the `row.rowPane` div is empty, so I suppose it's redrawn somehow and by that removing the active class?
- I add added some `div` wrappers for the TWBS card display, but updated `getTopTabContentHolder()` correspondingly, so I think it should work (and it does for any click after the initial render ;))

Maybe you have some ideas?

pmk65 commented:
For some reason the Travis test have stalled. And I can't see any way to restart it.
@stroebjo Try creating a dummy commit and see if it will trigger a  new test run.

schmunk42 commented:
Nice! Thank you!

pmk65 commented:
@stroebjo I'm creating a new issue where we can discuss the issues you have with the Bootstrap4 theme, like the questions you have posted above.  (Otherwise it get lost since the PR is now closed) 
So post a list of the remaining problems (just copy past it from this)

odurbalau commented:
Running the newest code in master with the schema below results in the following bug where list items appear outside of the unordered list (see attached image)
![image](https://user-images.githubusercontent.com/4391753/70480809-84484700-1a95-11ea-8535-9c15b291fa15.png)

```javascript
{
    "type": "object",
    "format": "categories",
    "properties": {
        "names": {
            "type": "array",
            "items": {
                "type": "object",
                "required": ["name"],
                "properties": {
                    "name": {"type": "string"}
                }
            }
        },
        "some_number": {"type": "number"}
    }
}
```
